tests/utils/state_locks_filter.py

```diff
@@ -0,0 +1,398 @@
+            print("")
+            print(trace)
+            print("")
+            for time in times:
+                print(
+                    "{} held: {} ms, wait: {} ms".format(
+                        time.get_group_name(),
+                        time.get_held_time(),
+                        time.get_wait_time(),
+                    )
+                )
+            print("")
+
+
+"""
+LocksFileReader: Reads the lock file and parses the different test cases and groups.
+It extracts the relevant trace data and stores it in LocksGroup instances.
+It can also return a dictionary of traces and associated group times.
+"""
+
+
+class LocksFileReader:
+
+    PROJECT_PREFIX = "###START: SNAPD PROJECT"
+    TEST_PREFIX = "###START:"
+
+    lines: list[str]
+    groups: list[LocksGroup]
+
+    def __init__(self, locks_file: argparse.FileType):
+        self.lines = []
+        self.groups = []
+
+        self._read(locks_file)
+
+    def _read(self, locks_file: argparse.FileType):
+        self.lines =  locks_file.readlines()
+
+        current_line = 0
+        if not self._is_project(self.lines[current_line]):
+            print("First time expected to be the project start.")
+            sys.exit(1)
+
+        # Read the tests
+        while True:
+            if current_line >= len(self.lines):
+                return
+
+            group_lines = self._current_group(current_line)
+            if len(group_lines) == 0:
+                raise RuntimeError("Error parsing test.")
+
+            self.groups.append(LocksGroup(group_lines))
+            current_line = current_line + len(group_lines)
+
+    # Indicates if the line is the project declaration
+    def _is_project(self, line: str) -> bool:
+        return line.startswith(self.PROJECT_PREFIX)
+
+    # This function works to detect the current test but also for the current project
+    def _current_group(self, start_line: int) -> list[str]:
+        next_match = get_next_match(self.lines, start_line, self.TEST_PREFIX)
+        if next_match == -1:
+            return self.lines[start_line:len(self.lines)]
+        return self.lines[start_line:next_match]
+
+    # Retrieve the test lines
+    def get_test(self, test: str) -> list[str]:
+        for group in self.groups:
+            if test in group.get_name():
+                return str(group)
+
+        return ""
+
+    # Retrieve the times for each trace in the file
+    # For each trace, there is a list with the times for each test where the trace appears
+    def get_traces_times(self) -> dict[LockOpTrace, list[GroupTimes]]:
+        traces = dict[LockOpTrace, list[GroupTimes]]()
+        for group in self.groups:
+            group_traces = group.get_traces()
+            for trace in group_traces:
+                group_time = GroupTimes(
+                    group.get_name(),
+                    group.get_lock_held_time(trace),
+                    group.get_lock_wait_time(trace),
```

Comment:
get_lock_held_time/get_lock_wait_time are almost identical but for each we iterate over all traces in the group and use:
```
           if lock.get_trace() == trace:
                return lock.get_wait_time()
```
which AFAIU will call to `__eq__(self, other)`,  which in turn calls `hash(self.__str__()) == hash(other.__str__())` where string representation joins all the lines of a trace. So we're computing the same `join()` over and over for every trace. It's probably acceptable as long as data set isn't too large, hopefully what we collect in the tests is rather small.

Reply:
I update that to calculate the hash when the trace is created, so that join is calculated just once

Reply:
The calculation is now done once but hash method on the type is never called due to another bug so the performance cost is back.